### PR TITLE
Move Dynniq EC-2 action to Windows

### PR DESCRIPTION
**Pull request by otterdahl**
The EC-2 runner has recently been moved to the Windows platform. This requires several updates to the workflow

- [x] Use Powershell instead of bash (or make WSL work?)
- [x] Make Ruby work

### CODE REVIEW

File: .github/workflows/dynniq_ec2.yml [hunk 1]
```diff
@@ -31,6 +31,7 @@ jobs:
       env:
         ImageOS: win19
       with:
+        ruby-version: 2.6
```
Comment (emiltin) [inline .github/workflows/dynniq_ec2.yml:34]: any reason not to use 3.0? that's what i'm running locally (on mac)

Comment (otterdahl) [reply .github/workflows/dynniq_ec2.yml:34]: No reason. Let's try 3.0

**Comment by emiltin** [comment]
There's an annoying error in JSONSchemer:

```
JSONSchemer::InvalidFileURI:
14
  cannot have a host (use `file:///`)
15
# ./vendor/bundle/ruby/2.6.0/gems/json_schemer-0.2.17/lib/json_schemer.rb:42:in `block in <module:JSONSchemer>'
16
# ./vendor/bundle/ruby/2.6.0/gems/json_schemer-0.2.17/lib/json_schemer/cached_ref_resolver.rb:10:in `call'
17
# ./vendor/bundle/ruby/2.6.0/gems/json_schemer-0.2.17/lib/json_schemer.rb:57:in `schema'
18
# ./vendor/bundle/ruby/2.6.0/bundler/gems/rsmp-9383b3296fb7/lib/rsmp/message.rb:16:in `load_schemas'
19
# ./vendor/bundle/ruby/2.6.0/bundler/gems/rsmp-9383b3296fb7/lib/rsmp/message.rb:30:in `<class:Message>'
20
# ./vendor/bundle/ruby/2.6.0/bundler/gems/rsmp-9383b3296fb7/lib/rsmp/message.rb:3:in `<module:RSMP>'
21
# ./vendor/bundle/ruby/2.6.0/bundler/gems/rsmp-9383b3296fb7/lib/rsmp/message.rb:2:in `<top (required)>'
22
# ./vendor/bundle/ruby/2.6.0/bundler/gems/rsmp-9383b3296fb7/lib/rsmp.rb:30:in `require'
23
# ./vendor/bundle/ruby/2.6.0/bundler/gems/rsmp-9383b3296fb7/lib/rsmp.rb:30:in `<top (required)>'
24
# ./spec/support/spec_helper.rb:2:in `require'
25
# ./spec/support/spec_helper.rb:2:in `<top (required)>'
```

That references these lines or Ruby code in Message#load_schemas:
```ruby
      core_schema_path = File.join(schema_path,'core','rsmp.json')
      @@schemas[nil] = JSONSchemer.schema( Pathname.new(core_schema_path) )
```

Seems to be an issuse with file paths on Windows?

**Comment by emiltin** [comment]
yay, ruby 3 running!
But why are we getting these file path errors only for ec-2, not itc-2? Must be a difference in the setup of these two?

**Comment by otterdahl** [comment]
> But why are we getting these file path errors only for ec-2, not itc-2? Must be a difference in the setup of these two?

The ec-2 runner is using in a windows environment where all the tools are compiled for windows. E.g. powershell instead of bash/zsh, git, 7zip, curl, ruby, etc.. There might some kind of problems with the paths. It requires further investigation

itc-2 uses a standard linux environment.

**Comment by emiltin** [comment]
You can try loading the schema files manually inside irb: 
```ruby
schema_path = File.join(File.dirname(__dir__),'rsmp_schema','schema')
core_schema_path = File.join(schema_path,'core','rsmp.json')
JSONSchemer.schema( Pathname.new(core_schema_path) )
```

__dir__ should be the current working directly.

**Comment by otterdahl** [comment]
Hmm. I'm not sure how to do this...

```powershell
PS C:\hostedtoolcache\windows\Ruby\3.0.0\x64\bin> .\irb
irb(main):001:0> schema_path = File.join(File.dirname(__dir__),'rsmp_schema','schema')
=> "./rsmp_schema/schema"
irb(main):002:0> core_schema_path = File.join(schema_path,'core','rsmp.json')
=> "./rsmp_schema/schema/core/rsmp.json"
irb(main):003:0> JSONSchemer.schema( Pathname.new(core_schema_path) )
Traceback (most recent call last):
        4: from C:/hostedtoolcache/windows/Ruby/3.0.0/x64/bin/irb.cmd:31:in `<main>'
        3: from C:/hostedtoolcache/windows/Ruby/3.0.0/x64/bin/irb.cmd:31:in `load'
        2: from C:/hostedtoolcache/windows/Ruby/3.0.0/x64/lib/ruby/gems/3.0.0/gems/irb-1.3.0/exe/irb:11:in `<top (required)>'
        1: from (irb):3:in `<main>'
NameError (uninitialized constant JSONSchemer)
irb(main):004:0>
```

**Comment by emiltin** [comment]
when in irb, you need to ```require 'json_schemer'```

**Comment by otterdahl** [comment]
```powershell
PS C:\hostedtoolcache\windows\Ruby\3.0.0\x64\bin> .\irb
irb(main):001:0> require 'json_schemer'
Traceback (most recent call last):
        6: from C:/hostedtoolcache/windows/Ruby/3.0.0/x64/bin/irb.cmd:31:in `<main>'
        5: from C:/hostedtoolcache/windows/Ruby/3.0.0/x64/bin/irb.cmd:31:in `load'
        4: from C:/hostedtoolcache/windows/Ruby/3.0.0/x64/lib/ruby/gems/3.0.0/gems/irb-1.3.0/exe/irb:11:in `<top (required)>'
        3: from (irb):1:in `<main>'
        2: from <internal:C:/hostedtoolcache/windows/Ruby/3.0.0/x64/lib/ruby/3.0.0/rubygems/core_ext/kernel_require.rb>:85:in `require'
        1: from <internal:C:/hostedtoolcache/windows/Ruby/3.0.0/x64/lib/ruby/3.0.0/rubygems/core_ext/kernel_require.rb>:85:in `require'
LoadError (cannot load such file -- json_schemer)
```

**Comment by emiltin** [comment]
their docs has this example:
```ruby
schema = Pathname.new('/path/to/schema.json')
schemer = JSONSchemer.schema(schema)
```

and we essentially do

```ruby
schema = Pathname.new('./rsmp_schema/schema/core/rsmp.json')
schemer = JSONSchemer.schema(schema)
```

Windows uses backslash for path separators, but Ruby stil uses '/'. Filepath will convert to the correct type when converting to strings. So I think our code should be ok.

This works for me in irb on mac:

```ruby
code/rsmp % cd lib
rsmp/lib % irb   
irb(main):001:0> require 'json_schemer'
=> true
irb(main):002:0> path=Pathname.new("./rsmp_schema/schema/core/rsmp.json")
=> #<Pathname:./rsmp_schema/schema/core/rsmp.json>
irb(main):003:0> uri = URI.parse("file://#{path.realpath}")
=> #<URI::File file:///Users/emiltin/Documents/code/rsmp/lib/rsmp_schema/schema/core/rsmp.json>
irb(main):004:0> uri.host && !uri.host.empty?
=> false
```

Note that because __dir__ is the folder containing the ruby code. You have to be inside the rsmp gem, in the lib folder.

**Comment by emiltin** [comment]
try to use 'bundler exec irb' inside the rsmp gem folder. that should ensure you have access to the gems.

**Comment by otterdahl** [comment]
```ruby
irb(main):001:0> require 'json_schemer'
=> true
irb(main):002:1* path=Pathname.new("./rsmp_schema/schema/core/rsmp.json")
=> #<Pathname:./rsmp_schema/schema/core/rsmp.json>
irb(main):003:0> uri = URI.parse("file://#{path.realpath}")
=> #<URI::File file://C/Users/DQ/actions-runner/_work/rsmp_validator/rsmp_validator/vendor/bundle/ruby/3.0.0/bundler/gems/rsmp-9383b3296fb...
irb(main):004:0> uri.host && !uri.host.empty?
=> true
```

**Comment by otterdahl** [comment]
If I change this line:
```ruby
uri = URI.parse("file:///#{path.realpath}")
```
to use 3 slashes instead. I get:
```ruby
irb(main):011:0> uri.host && !uri.host.empty?
=> false
```

**Comment by emiltin** [comment]
I'm trying a hack, to prepend a '/':
```ruby
path=Pathname.new("./rsmp_schema/schema/core/rsmp.json")
path.define_singleton_method(:realpath) { "/#{super()}" } # hack it
uri = URI.parse("file://#{path.realpath}")
```

**Comment by emiltin** [comment]
Seems like a bug in json_schemer. I've open en issue at https://github.com/davishmcclurg/json_schemer/issues/87

**Comment by otterdahl** [comment]
I've tested `rsmp_validator` on a different Windows 10 machine, running the very latest version of Windows (just like the EC-2 runner), where Ruby 3.0 is installed manually using the official [RubyInstaller](https://rubyinstaller.org/). It also produces the same error message.

**Comment by emiltin** [comment]
> I've tested `rsmp_validator` on a different Windows 10 machine, running the very latest version of Windows (just like the EC-2 runner), where Ruby 3.0 is installed manually using the official [RubyInstaller](https://rubyinstaller.org/). It also produces the same error message.

Yeah seems like a bug. Let's see if they respond to the issue i opened, otherwise we can fork their repo to fix it.

**Comment by emiltin** [comment]
I forked json_schemer and removed the offending check, which seems to work.

Now there's a different error:
```
Run bundle exec rspec spec/site

An error occurred while loading support/spec_helper.
Failure/Error: validator_config = YAML.load_file 'config/validator.yaml'

ArgumentError:
  ASCII incompatible encoding needs binmode
# ./spec/support/spec_helper.rb:55:in `<top (required)>'
No examples found.
```

Windows..

**Comment by emiltin** [comment]
The encoding error could be because the file config/validator.yaml was created on Windows, with an odd encoding?

**Comment by otterdahl** [comment]
> The encoding error could be because the file config/validator.yaml was created on Windows, with an odd encoding?

Yes. It's created as part of the workflow. It seems to be a quirk of powershell

**Comment by otterdahl** [comment]
> I forked json_schemer and removed the offending check, which seems to work.

Great work!

**Comment by emiltin** [comment]
seems to be humming along now!

**Comment by otterdahl** [comment]
> seems to be humming along now!

Yes. Looks good. Now I'm trying to find out why the EC-2 won't connect.

**Comment by emiltin** [comment]
The ec2 is takings its' sweet time, 27min and counting

**Comment by otterdahl** [comment]
> The ec2 is takings its' sweet time, 27min and counting

It failed to connect due to Windows built in firewall. I've added an exception. Let's try again.

**Comment by otterdahl** [comment]
I think the scope of this PL has be achieved. The EC-2 CI runner seems to working. Merge?

**Pull request merged**